### PR TITLE
Replace _strdup with instrumented ebpf_duplicate_string and fix bugs uncovered

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Description

Replace _strdup with ebpf_duplicate_string, which is instrumented to test low-memory conditions.
Fix bugs in _update_global_helpers_for_program_information uncovered by low-memory testing.

Resolves: #1699

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

**Review by gtrevi — APPROVED** [review]

File: libs/platform/ebpf_platform.h [hunk 1]
```diff
@@ -279,6 +279,15 @@ extern "C"
...
     ebpf_duplicate_utf8_string(_Out_ ebpf_utf8_string_t* destination, _In_ const ebpf_utf8_string_t* source);
 
+    /**
+     * @brief Duplicate a null-terminated string.
+     *
+     * @param[in] source String to duplicate.
+     * @return Pointer to the duplicated string or NULL if out of memory.
+     */
+    _Must_inspect_result_ char*
+    ebpf_duplicate_string(_In_ const char* source);
```
Comment (saxena-anurag) [inline libs/platform/ebpf_platform.h:289]:
```suggestion
    ebpf_duplicate_string(_In_z_ const char* source);
```

File: libs/platform/ebpf_object.c [hunk 2]
```diff
@@ -356,3 +356,15 @@ ebpf_object_reference_by_handle(
     return ebpf_reference_base_object_by_handle(
         handle, _ebpf_object_compare, &object_type, (ebpf_base_object_t**)object);
 }
+
+_Must_inspect_result_ char*
+ebpf_duplicate_string(_In_ const char* source)
```
Comment (saxena-anurag) [inline libs/platform/ebpf_object.c:361]:
```suggestion
ebpf_duplicate_string(_In_z_ const char* source)
```

File: libs/api/ebpf_api.cpp [hunk 3]
```diff
@@ -2105,7 +2115,12 @@ _ebpf_pe_add_section(
         EBPF_LOG_EXIT();
         return 1;
     }
-    info->program_type_name = _strdup(info->program_type_name);
+    info->program_type_name = ebpf_duplicate_string(info->program_type_name);
+    if (info->program_type_name == nullptr) {
+        pe_context->result = EBPF_NO_MEMORY;
+        EBPF_LOG_EXIT();
+        return 1;
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:2122]: Same.

File: libs/api/ebpf_api.cpp [hunk 4]
```diff
@@ -2084,47 +2086,77 @@ _ebpf_pe_add_section(
...
-        return 1;
+        return_value = 1;
+        goto Exit;
+    }
+
+    info->program_type_name = ebpf_duplicate_string(info->program_type_name);
+    if (info->program_type_name == nullptr) {
+        pe_context->result = EBPF_NO_MEMORY;
+        return_value = 1;
+        goto Exit;
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:2129]: There is a possibility of freeing a const string if `ebpf_duplicate_string` at line 2125 fails. I think we should save the result of `ebpf_get_program_type_name` (line 2118) in a local variable and then duplicate it in line 2125?

**Review by saxena-anurag — DISMISSED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Review by gtrevi — APPROVED** [review]

**Pull request merged**